### PR TITLE
Type promotion rule fix

### PR DESCRIPTION
**Pull request by shmsong**
I would like to use this PR for refining the type promotion rules among N-dim tensor, 0-dim tensor, and scalar inputs. 

Eager mode has a precedence rules among the three categories of inputs, defined in: 
[ATen/native/TypeProperties.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TypeProperties.cpp#L59
)

This short PR tries to replicate the same logic in applying type-promotion rules.

The python tests still need to be refactored to improve the coverage. I will push the refactor in a separate PR.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -251,11 +271,23 @@ DataType getOutputType(BinaryOpType op_type, Val* v1, Val* v2) {
...
+    v1_dtype = common_type;
+    v2_dtype = common_type;
+  }
+
   if (isIntegerOp(op_type) || (alsoBooleanOperator(op_type) && integer_input)) {
     // If integer op or maybe bool op with integer inputs meaning binary op
     if (integer_input && all_integer_input) {
       return promote_type(v1_dtype, v2_dtype);
     } else if (integer_input && !all_integer_input) {
+      // Question: this doesn't look like what eager mode does
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: @csarofeen  I actually have a question on this logic since eager mode doesn't seem to do this, for example: 
```python
a=torch.tensor([1],dtype=torch.float,device="cuda")
b=torch.tensor([1],dtype=torch.long,device="cuda")
c=a>>b # c.dtype == torch.float
```
I was wondering if it is safe to change this part.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: I don't think we should any bitwise operators that would return float tensors at this time. I'm uncertain who should use them or why.

File: test/test_jit_cuda_fuser.py [hunk 2]
```diff
@@ -418,10 +454,11 @@ def t(x: torch.Tensor):
             return o
 
         try:
-            if data is None:
-                x = torch.randn(shape, dtype=dtype, device="cuda")
+            if random_data:
+                x = torch.randn(shape, dtype=torch.float32, device="cuda")
+                x = x.to(dtype=dtype)
```
Comment (csarofeen) [inline test/test_jit_cuda_fuser.py]: If it's an int type can we multiply by something like 5 to get a larger variance?

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: Sure. Added assertion here so we just fallback if any user happens to use this kind of op.

File: test/test_jit_cuda_fuser.py (hunk 2 above)
Comment (shmsong) [reply test/test_jit_cuda_fuser.py]: Updated. Thanks!
